engine/src/main/java/org/hibernate/validator/engine/HibernateConstraintViolation.java

```diff
@@ -0,0 +1,21 @@
+/*
+ * Hibernate Validator, declare and validate application constraints
+ *
+ * License: Apache License, Version 2.0
+ * See the license.txt file in the root directory or <http://www.apache.org/licenses/LICENSE-2.0>.
+ */
+package org.hibernate.validator.engine;
+
+import javax.validation.ConstraintViolation;
+
+/**
+ * A custom {@link ConstraintViolation} which allows to get additional information for a constraint violation.
+ *
+ * @since 5.3
+ */
+public interface HibernateConstraintViolation<T> extends ConstraintViolation<T> {
+	/**
+	 * @return an object that further describes the violation or {@code null}.
+	 */
+	Object getViolationContext();
```

Comment:
In the beginning I thought about this too and actually even wanted to go one step further so that the user defines the violation context type on the custom constraint annotation.

This type could then have even been used to do some runtime type validation, when it is tried to set a violation context of a different type.

```
@Target( { METHOD, FIELD, ANNOTATION_TYPE, TYPE, PARAMETER })
@Retention(RetentionPolicy.RUNTIME)
@Constraint(validatedBy = { FooValidator.class })
@Documented
public @interface FooConstraint {
	String message() default "";
	Class<?>[] groups() default {};
	Class<? extends Payload>[] payload() default {};
	Class<?>[] violationContext() default {};
}
```

Also from your proposal, it would then make sense to throw an exception when the type parameter given differs from what have been specified in the annotation.

WDYT?

However we can start introducing the definition you suggested and make it return null in case there is a violation context but of a different type.